File: src/ginger/hardware/infrared.py
"""3-sensor infrared line tracker.

Sensors are active-low: value=1 means line detected (dark surface).
GPIO pins: left=14, center=15, right=23.
"""

from gpiozero import LineSensor

_PINS = {"left": 14, "center": 15, "right": 23}


class InfraredSensors:
    def __init__(self):
        self._sensors = {name: LineSensor(pin) for name, pin in _PINS.items()}

    def read(self, sensor: str) -> bool:
        """True = line detected under this sensor."""
        return bool(self._sensors[sensor].value)

    def read_all(self) -> dict[str, bool]:
        return {name: bool(s.value) for name, s in self._sensors.items()}

    def close(self) -> None:
        for s in self._sensors.values():
            s.close()

File: ginger-rs/src/infrared.rs
//! 3-sensor infrared line tracker.
//!
//! GPIO: left=14, center=15, right=23. Active-low: High = line detected.

use rppal::gpio::{Gpio, InputPin, Level};

use crate::Result;

const PIN_LEFT:   u8 = 14;
const PIN_CENTER: u8 = 15;
const PIN_RIGHT:  u8 = 23;

pub struct InfraredSensors {
    left:   InputPin,
    center: InputPin,
    right:  InputPin,
}

impl InfraredSensors {
    pub fn new() -> Result<Self> {
        let gpio = Gpio::new()?;
        Ok(Self {
            left:   gpio.get(PIN_LEFT)?.into_input(),
            center: gpio.get(PIN_CENTER)?.into_input(),
            right:  gpio.get(PIN_RIGHT)?.into_input(),
        })
    }

    /// Returns (left, center, right): true = line detected.
    pub fn read_all(&self) -> (bool, bool, bool) {
        (
            self.left.read()   == Level::High,
            self.center.read() == Level::High,
            self.right.read()  == Level::High,
        )
    }
}
